evm/src/propeller-router/PropellerRouter.sol

```diff
@@ -529,4 +537,78 @@ contract PropellerRouter is
             amount == 0 ? _weth.balanceOf(address(this)) : amount;
         _weth.withdraw(unwrapAmount);
     }
+
+    /**
+     * @dev Check if the sender is correct and executes callback actions.
+     *  @param msgData encoded data. It must includes data for the verification and the action.
+     */
+    function _executeGenericCallback(bytes calldata msgData) internal {
+        (
+            uint256 amountOwed,
+            uint256 amountReceived,
+            address tokenOwed,
+            uint16 offset
+        ) = _verifyCallback(msgData);
+
+        bytes calldata data = msgData[4 + offset:];
+        ActionType supplyActionType;
+        ActionType forwardActionType;
+        bytes calldata supplyActionData;
+        bytes calldata forwardActionData;
+        assembly {
+            // 1 + 1 + 4 = 6
+            let dataLoad := calldataload(data.offset)
+            supplyActionType := and(shr(248, dataLoad), 0xff)
+            forwardActionType := and(shr(240, dataLoad), 0xff)
+            // 2 bytes offset from header containing actiontypes + 2 + 2 bytes of length encoding
+            supplyActionData.offset := add(data.offset, 10)
+            supplyActionData.length := and(shr(224, dataLoad), 0xffff)
+            // any remaining bytes are calldata for forwardReceived
+            forwardActionData.offset :=
+                add(supplyActionData.offset, supplyActionData.length)
+            forwardActionData.length := and(shr(192, dataLoad), 0xffff)
+        }
+
+        _swapCallback(
+            IERC20(tokenOwed),
+            amountOwed,
+            amountReceived,
+            supplyActionType,
+            supplyActionData,
+            forwardActionType,
+            forwardActionData
+        );
+    }
+
+    /**
+     * @dev Generic swap callback, can perform other actions see _executeAction.
+     */
+    function _swapCallback(
+        IERC20 tokenOwed,
+        uint256 amountOwed,
+        uint256 amountReceived,
+        ActionType supplyOwedActionType,
+        bytes calldata actionData0,
+        ActionType forwardReceivedActionType,
+        bytes calldata actionData1
+    ) internal {
+        // msg.sender must be a pool this needs to be verfied before calling this method!
+        if (supplyOwedActionType == IBatchSwapRouterV1Structs.ActionType.PAY) {
+            address payer = _getPayer();
+            _resetPayer();
+            tokenOwed.safeTransferFrom(payer, msg.sender, amountOwed);
```

Comment:
I thought we said we wouldn't do this.. that we would only do transfer and that simplifies the callbacks

Reply:
i literally copy-pasted this from the last router. I'd like to implement this totally from scratch and just use this code snippet afterwards to make sure we didn't miss anything. Didn't get around to it though. Too much to do, so little time